### ISSUE TITLE
Remove C++ Bindings

### ISSUE THREAD

**Issue by mpiforumbot** [mpi-forum/mpi-forum-historic#281]
**Originally by _ftillier_ on 2011-06-23 17:13:11 -0500**

---

This ticket removes the deprecated C++ bindings.
## Motivation

The C++ bindings are problematic for binary distributions of implementations because there is no standard for C++ name mangling.  This requires implementors to either ship multiple binaries (one set of binaries per target compiler), restrict customer choice in compiler suites, or ship the C++ bindings in source form, requiring them to be layered above the C bindings.

Further, new MPI-3.0 routines have no C++ bindings, requiring C++ applications to use the C bindings.
## Backward Compatibility

Applications requiring support for these can request support for them from their MPI implementations. There are no problems with concurrent support of different versions of the standard, allowing implementations to be compliant with both MPI 2.2 as well as MPI 3.0.  The C++ bingings can thus be provided via MPI 2.2 support, while new MPI-3.0 specific functionality can be used via the C bindings.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2011-07-20 20:41:12 -0500**

---

As expected, there was much discussion about this at the Forum.

The usual points were brought up:
- There are existing apps out there that use the C++ bindings (e.g., DOE apps and Intel has some customers that use them)
- Rolf cited the MPI survey:
  - 165 respondents said they have C++ apps that use the MPI C++ bindings
  - 550 said they didn't have C++ apps that use the MPI C++ bindings
  - 107 said they didn't know if they have C++ apps that use the MPI C++ bindings
- There was discussion that it would be good to fix the dichotomy of having C++ for some MPI API functions, but not all (i.e., not new 3.0 API functions) either by removing the C++ bindings or undeprecating them.
- There is an ancient ticket to undeprecate the C++ bindings that has not moved in over a year.
- Surprisingly (to me, anyway), some DOE personnel said that the current state of deprecated C++ was perfect for them:
  - Their existing C++ MPI applications will just use the C bindings for new MPI-3 functionality (and those app developers are ok with that)
  - New MPI applications will not use the C++ MPI API
  - Hence, deprecated is _perfect_ for them -- they don't want to undeprecate, and they don't want to remove.
- I took 2 straw polls.  Poll !#1:
  - Delete the C++ bindings (probably moving them off to a "deleted" chapter, per the idea raised in #279): 8
  - Keep the C++ bindings as they are now (i.e., deprecated): 4
  - Abstain: 8
- Straw poll !#2:
  - Undeprecate the C++ bindings: 5
  - Keep the C++ bindings deprecated: 10
  - Abstain: 6
- These are not conclusive straw polls because of a lot of abstains, but at least half of the attendees want to keep the C++ bindings deprecated, and 2/5 want to delete the C++ bindings altogether.

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-09-01 13:04:35 -0500**

---

Withdrawing this ticket, as ticket #279 makes the C++ bindings optional and is far less contentious, as well as a lot easier in terms of LaTeX changes.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2011-10-25 14:00:51 -0500**

---

accepting as potential work to enumerate chagens required to delete C++ bindings

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2011-11-02 08:59:27 -0500**

---

Added a first-pass at removing the C++ bindings. I created a new chapter "Removed Language Interfaces" and moved all C++ function prototypes there, as well as the design discussion and listings of Language Constants.

In most cases, the original text is marked with "ticket281" and the removed C++ text colored as DELETE. There were some impracticalities so tables with removed C++ columns and some "trivial" references to things like "C/C++" are not marked, however I have revision history in my repository to track those.

I'm not sure I captured the original intent for "removing C++ bindings", but it is a start. What's the next step?

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-11-02 10:04:36 -0500**

---

I'd suggest naming the new chapter "Removed Interfaces" to not make it programming language specific.  For example, there is a ticket I started, to remove some of the deprecated MPI-1 routines.  For these, all bindings would be removed, and it would be nice if this new chapter could capture these too.

I think the language should be faily brief in the deleted chapter.  It should reference the names of the objects/APIs/constants/etc, say they were deleted, and just reference the last standard that included them.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2011-11-29 14:00:07 -0600**

---

Updated svn with new name for removed interfaces section. attached new pdf.

**Comment by mpiforumbot** [comment]
**Originally by _ftillier_ on 2011-12-02 13:05:37 -0600**

---

Hi Doug,

I've reviewed the document, and it looks good overall.  Here's the few comments I came up with:
Page 15, line 15: No need for a blank line with a period
Page 16, line 10: Should still say "In C one uses MPI_Offset."
Page 514, Line 38: The footnote should go away.
Page 518, probably a LaTeX issue, but it does not show as new text.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-06 13:23:55 -0600**

---

Doug -- a few comments on the PDF.
- p16: removes "In C one uses MPI_Offset" -- don't we still need this?
- p490: s/Chatper/Chapter/
- p491: s/may only be provided/may be provided/
- p491: WRT: "Where an MPI-3.0 interface does not have a corresponding C++ interface,", no MPI-_3.0_ interfaces will have C++ _bindings_
- p514: caption on "Null Handles" table needs to change
- p539: "moved entire section on C and C++ interoperability into REMOVED section(s)" -- where?  I don't see it in chapter 16

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-06 13:40:23 -0600**

---

Replying to jsquyres:

> Doug -- a few comments on the PDF.
> - p16: removes "In C one uses MPI_Offset" -- don't we still need this?
>   I took this sentence to be a comparison between C and C++. I don't think it makes any sense to just say "In C one uses MPI_Offset." since it is an example on how to get a C++ name from a C name.
> - p490: s/Chatper/Chapter/
>   Fixed.
> - p491: s/may only be provided/may be provided/
>   I don't know for sure where that text originated, but since it doesn't pertain to C++ bindings I did not examine it.  I think that sort of change is for another ticket?
> - p491: WRT: "Where an MPI-3.0 interface does not have a corresponding C++ interface,", no MPI-_3.0_ interfaces will have C++ _bindings_
>   Removed text.
> - p514: caption on "Null Handles" table needs to change
>   This table was simply moved from the "Constants" section and modified to refer only to C++ names. Should both tables (C/FORTRAN and Removed C++ sections) have their heading changed? I'm not sure that's within the scope of this ticket.
> - p539: "moved entire section on C and C++ interoperability into REMOVED section(s)" -- where?  I don't see it in chapter 16
>   That section was originally moved to the Removed Interfaces section, but later I got feedback saying it should be deleted. I'll change this note accordingly.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-06 14:58:20 -0600**

---

updated attached document with "ticket0" changes from Jeff Squyres.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-01-10 22:10:08 -0600**

---

Proposal was read in the Jan 2012 Forum meeting.  _Even more_ textual cuts were proposed by the Forum.  Additionally, several C++ sections were missed that need to be cut.  

So the proposal will be revised a bit more and brought up again for a reading at the next meeting.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-11 16:47:56 -0600**

---

**Attachment added:** [`mpi3-ticket281-report.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/281/mpi3-ticket281-report.pdf) (3677.0 KiB)
MPI3 preliminary report with C++ bindings moved to REMOVED section

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-01-11 16:48:39 -0600**

---

Updated proposal doc to eliminate C++ refs that crept back in due to macros, etc.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-03-06 11:40:25 -0600**

---

**Attachment added:** [`ticket281.diffs`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/281/ticket281.diffs) (193.2 KiB)
Diffs since last forum reading

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-03-06 11:42:16 -0600**

---

**Attachment added:** [`ticket281.patch`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/281/ticket281.patch) (193.2 KiB)
Diffs since last forum reading

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-01 04:30:09 -0500**

---

There was an email-discussion on "C++ types inaccessible after #281" on
the Forum list. This is now continued in the MPI-2.2 errata ticket #340,
which tries to solve the problem how to support special C++ types in the future
when the MPI C binding is used for C++ application codes.

**Comment by mpiforumbot** [comment]
**Originally by _davesolt_ on 2012-07-10 15:59:34 -0500**

---

On page 324 line 13 of the mpi3-ticket281-report.pdf it says "In C++ this constant does not exist, and the array_of_errcodes argument may be omitted from the argument list".   I see many places in the document where similar references to the C++ interface exist.

Page 383 it says "and in C++", but then has nothing afterwards.  This same issue is repeated in several places. 

There may be other locations.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2012-07-12 16:19:13 -0500**

---

I looked at this ticket to prepare the merging if it gets accepted. I used the attached pdf because the patch and diff include tons of unrelated diffs. The pdf is in a completely inconsistent state so that I could not merge it. I started a review and found:

Page 243: dangling C++
Page 244: dangling C++
Page 247: 2x dangling C++
Page 248: dangling C++
Page 249: 2x dangling C++
Page 250: dangling C++
Page 254: dangling C++
Page 255: dangling C++

and more: 267 (2x, complete sentence), 269 (2x, complete sentence), 283, 285, 287 
(2x), 289, 290 (whole paragraph), 292, 293, 298 (2x), 299 (2x)

I stopped in page 299 deciding that this ticket is _not_ ready for vote or review. I am not sure what the folks who voted for this ticket reviewed, but I am seriously concerned about the quality of the ticket and the standard after the merge. 

I also checked "ticket281.patch" for topol.tex (Topology Chapter) and several references to C++ are not removed. 

Can somebody please clarify? 

Torsten

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-07-13 07:22:51 -0500**

---

When you say "dangling C++" are you talking about the chunk of text that is wrapped in an MPIdelete macro (i.e. surrounded by brackets and in color purple/magenta?). The only other C++ I see on page 243 is one that makes reference to programming languages that could be used, and it should remain.

What was your criteria for instances of "C++" that should or should not remain in the document?

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2012-07-13 09:07:15 -0500**

---

Replying to dougmill:

> When you say "dangling C++" are you talking about the chunk of text that is wrapped in an MPIdelete macro (i.e. surrounded by brackets and in color purple/magenta?). The only other C++ I see on page 243 is one that makes reference to programming languages that could be used, and it should remain.
> I am talking about the untouched parts that would not make much sense when the remaining C++ bindings are removed, i.e., the _not_ removed C++ stuff.

If we do not specify C++ bindings then we cannot specify constants for C++ (this is what the text on that page does). This is inconsistent, constant names are parts of bindings. The only consistent choice imho would be to remove all C++ from the document. 

Or see for example the _not removed_ statement on page 267 (topology chapter): "In C++, this constant does not exist and the weight arguments may be omitted from the argument list.". This doesn't make any sens with the removed C++ bindings. There are dozens more of those and the standard would be in a completely inconsistent state if we would pass this ticket!

> What was your criteria for instances of "C++" that should or should not remain in the document?
> If we remove language bindings, we should probably remove all of it because we don't specify for this language. We had this long discussion that C++ is mostly backwards-compatible to C, so specifying C would be all there is needed. Otherwise, we shall also start to include Java and Python in all discussions since there are third-party bindings for those (actually _more_ and _better_ ones than (complete) boost::mpi which was proposed as the replacement for the C++ bindings even though it only implements "the majority of functionality in MPI 1.1" (is missing 2.x and 3.x and there is no path forward for those).

Torsten

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-07-13 09:15:44 -0500**

---

This was discussed during the review. It was decided that these references were to programming languages, not bindings. The sentence states that the constant may be used with all supported programming languages: C, C++, and Fortran.  If we want to flip-flop back on this, the forum needs to make a conscious decision, again.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2012-07-13 11:05:53 -0500**

---

Replying to dougmill:

> This was discussed during the review. It was decided that these references were to programming languages, not bindings. The sentence states that the constant may be used with all supported programming languages: C, C++, and Fortran.  If we want to flip-flop back on this, the forum needs to make a conscious decision, again.
> I may have been in Washington during this discussion and there was no documentation of this decision at the ticket or anywhere else. I would have opposed immediately for the above mentioned reasons.

And clearly, this does not apply to the sentence "In C++, this constant does not exist and the weight arguments may be omitted from the argument list.". There are more of those issues. 

Just for fun, I invested two more minutes and found the following sentence remaining after applying this ticket (page 287) "The C and C++ bindings of MPI_ALLOC_MEM and MPI_FREE_MEM
are similar to the bindings for the malloc and free C library calls". Great!

I strongly object to this ticket, not only from a technical perspective but also from the perspective of the numerous inconsistencies it introduces to the chapters I am responsible for.

Torsten

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-07-13 14:07:50 -0500**

---

FYI, the changes I've made (so far) are in the SVN repository on a branch at:

```
trunk/working-groups/mpi-3/bw-compat/ticket-281
```

**Comment by mpiforumbot** [comment]
**Originally by _davesolt_ on 2012-07-16 11:48:00 -0500**

---

**Attachment added:** [`mpi-report_281_7_16_2012.pdf`](https://github.com/mpiforumbot/mpi-forum-historic/raw/master/attachments/281/mpi-report_281_7_16_2012.pdf) (2165.0 KiB)

**Comment by mpiforumbot** [comment]
**Originally by _anhvo_ on 2012-07-16 11:52:30 -0500**

---

The latest pdf built from the ticket-281 directory (posted above by Dave) seems to have addressed all the issues.

**Comment by mpiforumbot** [comment]
**Originally by _davesolt_ on 2012-07-16 11:56:15 -0500**

---

I attached the latest pdf of the ticket-281 branch.  I think that some of the confusion here is that the pdf posted to this ticket is very old and does not include a large round of edits done around April of 2012.   I do not see any problems in this latest pdf.

I do think that the forum failed to handle the voting of this ticket properly.  The voting of this ticket in Japan should have included looking at the actual textual changes and not simply a vote on the idea of removing the C++ bindings.  The document edits are quite significant for this ticket.  That said, I do think the actual latex changes done in April did cover the issues that Torsten and I brought up recently.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-17 05:25:14 -0500**

---

Review of mpi-report_281_7_16_2012.pdf
- p494:18 "(in C)" must not be removed.
- many locations where the macros are showing a wrong text in the removed part.
- It looks like that nearly all macro-usage must be checked about introducing 
  double spaces in the final doc:
  
    xxx \MPIdelete{....} yyy
  
  will produce 
  
    xxx**<SPACE><SPACE>**yyy.
  
  One has to choose
  
    xxx\MPIdelete{....} yyy 
  
  or
  
    xxx \MPIdelete{....}yyy
  
  which will produce
  
    xxx**<SPACE>**yyy

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-07-17 08:09:48 -0500**

---

Can you elaborate on your second point.  What do you mean by "showing a wrong text in the removed part"? 

There are places where "C++ macros" also had text fragments associated with them, and the macros are turned off so the deleted text does not read the same as it did before removal since the macro text is absent.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2012-07-17 14:27:37 -0500**

---

Committed to approved/topol, no changes for colls and one sided chapters. Please review!

--> http://www.unixer.de/sec/mpi-report.pdf

Torsten

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-17 15:18:52 -0500**

---

Committed changes to the bindings chapter.

Also disabled the \mpicppbind and \mpiemptycppbind macros, thereby effectively removing all C++ bindings from the document.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-18 06:53:10 -0500**

---

Passed July 17, 2012 with 15 YES, 4 NO, 2 ABSTAIN.

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-18 07:04:02 -0500**

---

The sources for this ticket, i.e.,
- tha attached latest pdf
- and the source on svn/mpi-forum-docs/trunk/working-groups/mpi-3/bw-compat/ticket-281/
  **may be** in a **terrible state**. (If this is only one location, then it is now detected.)

There are part(s) that are **deleted from the source without any notification** (e.g. macro or some comments telling else that this is a change due to #281).

A **grep 281 _/_.tex** does not show all modified locations!

Example:
- MPI-2.2 A.1.3, the C++ prototypes section, MPI-2.2 p527:42 - p528:32

I hope, that this is the only such modification without notification.

We need that this is clarified by the ticket responsibles, otherwise a correct implementation in the approved latex source is hard.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-18 07:12:05 -0500**

---

What you're saying is that the 281 authors didn't consistently use \MPIdelete and whatnot.  That is correct.

I actually went through a PDF built from that tree and manually added the relevant \MPIdelete (etc.) in the bindings chapter.  It was a little annoying, but I think I found all the Right places where 281 removed text.

**Comment by mpiforumbot** [comment]
**Originally by _dougmill_ on 2012-07-18 07:22:35 -0500**

---

This was discussed, at least at all the readings I did. No objections were raised at the time. That branch was created with initial, un-modified, files from the trunk at that time. Diffing between the initial rev and current rev should show all changes made.

So, are we keeping the \MPIdelete, etc, macros for the "proof" copy of the doc?  I was under the impression that the doc would be unreadable with all the annotations from all tickets kept in it.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-18 07:33:23 -0500**

---

Yes, the macros serve multiple purposes:
- Making it easy for chapter authors to find where all the changes were made
- Making it possible to proof exactly what was removed / changed / etc.

Since they're macros, it's easy to produce proofing copies for internal use in the forum (i.e., showing all the deleted text, etc.), but then also with the flick of a switch, _not_ show all the deleted text (etc.) and generate a final, clean copy for production that _doesn't_ include all the deleted text (etc).

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-18 07:44:15 -0500**

---

Replying to dougmill:

> ... Diffing between the initial rev and current rev should show all changes made.

A svn diff between which svn releases of 
svn/mpi-forum-docs/trunk/working-groups/mpi-3/bw-compat/ticket-281/
must be done to catch exactly the changes you did?

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-18 07:50:53 -0500**

---

Yes.  For example:

```
svn diff -r 969:HEAD https://svn.mpi-forum.org/svn/mpi-forum-docs/trunk/working-groups/mpi-3/bw-compat/ticket-281/chap-binding
```

shows the cumulative diff in the bindings chapter of all the changes made on the 281 branch.  Edit the URL to show whichever chapter you want, or remove the chapter directory to show _all_ changes.

**Comment by mpiforumbot** [comment]
**Originally by _htor_ on 2012-07-18 12:42:18 -0500**

---

Small updates to coll chapter based on Howard's comment (thanks).

--> http://www.unixer.de/sec/mpi-report.pdf

Thanks,
   Torsten

**Comment by mpiforumbot** [comment]
**Originally by _buntinas_ on 2012-07-18 14:46:13 -0500**

---

Applied to Info chapter in r1422.

-d

**Comment by mpiforumbot** [comment]
**Originally by _schulzm_ on 2012-07-18 16:02:24 -0500**

---

Applied to Environment Chapter 8

**Comment by mpiforumbot** [comment]
**Originally by _RolfRabenseifner_ on 2012-07-18 18:19:09 -0500**

---

Committed to approved/chap-appLang as r1440

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-18 18:24:01 -0500**

---

The removed chapter reviewed ok.

**Comment by mpiforumbot** [comment]
**Originally by _jsquyres_ on 2012-07-18 18:24:38 -0500**

---

All text appears to have been committed; moving to the "Waiting for PDF reviews" state.

**Comment by mpiforumbot** [comment]
**Originally by _buntinas_ on 2012-07-18 18:37:40 -0500**

---

Topo chapter reviewed and fixed.  OK

**Comment by mpiforumbot** [comment]
**Originally by _anhvo_ on 2012-07-18 18:42:48 -0500**

---

pt2pt reviewed. OK

**Comment by mpiforumbot** [comment]
**Originally by _buntinas_ on 2012-07-18 19:21:58 -0500**

---

Bindings chapter reviewed and fixed.  OK

**Comment by mpiforumbot** [comment]
**Originally by _anhvo_ on 2012-07-19 12:02:30 -0500**

---

In clean doc R1473 I'm still seeing a couple C++ bindings in chapter 13 (I/O). Page 541 and 542 (MPI_Datarep_extent_function and MPI_Datarep_conversion_function)

**Comment by mpiforumbot** [comment]
**Originally by _rsthakur_ on 2012-07-19 13:07:30 -0500**

---

Also disabled the \mpicpptypedefbind and \mpicpptypedefemptybind macros to remove the above C++ typedefs in the I/O chapter
